src/librustc_codegen_llvm/intrinsic.rs

```diff
@@ -143,14 +143,31 @@ impl IntrinsicCallMethods<'tcx> for Builder<'a, 'll, 'tcx> {
                 self.va_end(args[0].immediate())
             }
             "va_copy" => {
-                let va_list = match (tcx.lang_items().va_list(), &result.layout.ty.sty) {
-                    (Some(did), ty::Adt(def, _)) if def.did == did => args[0].immediate(),
-                    (Some(_), _)  => self.load(args[0].immediate(),
-                                               tcx.data_layout.pointer_align.abi),
-                    (None, _) => bug!("`va_list` language item must be defined")
+                let inner_ty = match tcx.lang_items().va_list() {
+                    Some(did) => tcx.type_of(did),
+                    None => bug!("`va_list` language item must be defined")
                 };
+                let impl_ty = match inner_ty.sty {
+                    ty::Adt(adt, _) => tcx.type_of(adt.non_enum_variant().fields[0].did),
+                    _ => bug!("improperly constructed `va_list` lang item")
+                };
+                // If the underlying structure `VaListImpl` is a structure,
+                // we need to allocate it (e.g., X86_64 on Linux).
                 let intrinsic = self.cx().get_intrinsic(&("llvm.va_copy"));
-                self.call(intrinsic, &[llresult, va_list], None);
+                if let ty::Adt(..) = impl_ty.sty {
+                    // Create an unnamed allocation for the backing structure
+                    // and store it in the the spoofed `VaList`.
+                    let layout = self.layout_of(inner_ty);
+                    let backing = PlaceRef::alloca(self, layout, "");
+                    let src = self.load(args[0].immediate(),
```

Comment:
This looks flipped. In the prior version we `load` for non-pointer variants and don't `load` for pointer variants.

Reply:
`impl_ty` being a `ty::Adt` means we're in a struct variant; for pointer variants, `impl_ty` is a `ty::Foreign`. Both `inner_ty` and `impl_ty` are retrieved from `va_list()`, so `impl_ty` should always match the definition of `VaListImpl`.

Reply:
In other words, `ty::Adt` for `VaListImpl` => struct variant => we want a load. Otherwise, `ty::Foreign` => pointer variant => no load. Did I get this wrong?

Reply:
Ah right, this was flipped in the previous version, but that was due to the old signature of `va_copy`.